// polywrap/build/lib/defaults/deploy-modules/local-dev-ens/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_js_1 = require("@polywrap/core-js");
var test_env_js_1 = require("@polywrap/test-env-js");
var path_1 = __importDefault(require("path"));
var client_js_1 = require("@polywrap/client-js");
var ethereum_plugin_js_1 = require("@polywrap/ethereum-plugin-js");
var contentHash = require("content-hash");
var LocalDevENSPublisher = /** @class */ (function () {
    function LocalDevENSPublisher() {
    }
    LocalDevENSPublisher.prototype.execute = function (uri, config) {
        return __awaiter(this, void 0, void 0, function () {
            var cid, ethereumPluginUri, client, signer, ensWrapperUri, _a, registerData, error, hash, setContenthashData;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (uri.authority !== "ipfs") {
                            throw new Error("ENS Deployer: resolved URI from " + uri + " does not represent an IPFS contentHash");
                        }
                        cid = uri.path;
                        ethereumPluginUri = "wrap://ens/ethereum.polywrap.eth";
                        client = new client_js_1.PolywrapClient({
                            plugins: [
                                {
                                    uri: ethereumPluginUri,
                                    plugin: ethereum_plugin_js_1.ethereumPlugin({
                                        networks: {
                                            testnet: {
                                                provider: "http://localhost:" + config.ports.ethereum,
                                            },
                                        },
                                        defaultNetwork: "testnet",
                                    }),
                                },
                            ],
                        });
                        return [4 /*yield*/, client.invoke({
                                method: "getSignerAddress",
                                uri: ethereumPluginUri,
                                input: {
                                    connection: {
                                        networkNameOrChainId: "testnet",
                                    },
                                },
                            })];
                    case 1:
                        signer = (_b.sent()).data;
                        if (!signer) {
                            throw new Error("Could not get signer");
                        }
                        ensWrapperUri = "fs/" + path_1.default.join(path_1.default.dirname(require.resolve("@polywrap/test-env-js")), "wrappers", "ens");
                        return [4 /*yield*/, client.invoke({
                                method: "registerDomainAndSubdomainsRecursively",
                                uri: ensWrapperUri,
                                input: {
                                    domain: config.domainName,
                                    owner: signer,
                                    resolverAddress: test_env_js_1.ensAddresses.resolverAddress,
                                    ttl: "0",
                                    registrarAddress: test_env_js_1.ensAddresses.registrarAddress,
                                    registryAddress: test_env_js_1.ensAddresses.ensAddress,
                                    connection: {
                                        networkNameOrChainId: "testnet",
                                    },
                                },
                            })];
                    case 2:
                        _a = _b.sent(), registerData = _a.data, error = _a.error;
                        if (!registerData) {
                            throw new Error("Could not register domain '" + config.domainName + "'" +
                                (error ? "\nError: " + error.message : ""));
                        }
                        return [4 /*yield*/, client.invoke({
                                method: "awaitTransaction",
                                uri: ethereumPluginUri,
                                input: {
                                    txHash: registerData.hash,
                                    confirmations: 1,
                                    timeout: 15000,
                                    connection: {
                                        networkNameOrChainId: "testnet",
                                    },
                                },
                            })];
                    case 3:
                        _b.sent();
                        hash = "0x" + contentHash.fromIpfs(cid);
                        return [4 /*yield*/, client.invoke({
                                method: "setContentHash",
                                uri: ensWrapperUri,
                                input: {
                                    domain: config.domainName,
                                    cid: hash,
                                    resolverAddress: test_env_js_1.ensAddresses.resolverAddress,
                                    connection: {
                                        networkNameOrChainId: "testnet",
                                    },
                                },
                            })];
                    case 4:
                        setContenthashData = (_b.sent()).data;
                        if (!setContenthashData) {
                            throw new Error("Could not set contentHash for '" + config.domainName + "'");
                        }
                        return [4 /*yield*/, client.invoke({
                                method: "awaitTransaction",
                                uri: ethereumPluginUri,
                                input: {
                                    txHash: setContenthashData.hash,
                                    confirmations: 1,
                                    timeout: 15000,
                                    connection: {
                                        networkNameOrChainId: "testnet",
                                    },
                                },
                            })];
                    case 5:
                        _b.sent();
                        return [2 /*return*/, new core_js_1.Uri("ens/" + config.domainName)];
                }
            });
        });
    };
    return LocalDevENSPublisher;
}());
exports.default = new LocalDevENSPublisher();
