// @polywrap/schema-compose/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveImports = exports.composeSchema = exports.ComposerFilter = exports.renderSchema = void 0;
var resolve_1 = require("./resolve");
var render_1 = require("./render");
Object.defineProperty(exports, "renderSchema", { enumerable: true, get: function () { return render_1.renderSchema; } });
var env_1 = require("./env");
var schema_parse_1 = require("@polywrap/schema-parse");
__exportStar(require("./types"), exports);
var ComposerFilter;
(function (ComposerFilter) {
    ComposerFilter[ComposerFilter["Schema"] = 1] = "Schema";
    ComposerFilter[ComposerFilter["TypeInfo"] = 2] = "TypeInfo";
    ComposerFilter[ComposerFilter["All"] = 3] = "All";
})(ComposerFilter = exports.ComposerFilter || (exports.ComposerFilter = {}));
function composeSchema(options) {
    return __awaiter(this, void 0, void 0, function () {
        var typeInfos, typeInfo, includeSchema, includeTypeInfo;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, resolveImports(options.schemas, options.resolvers)];
                case 1:
                    typeInfos = _a.sent();
                    typeInfo = typeInfos.length === 1 ? typeInfos[0] : schema_parse_1.combineTypeInfo(typeInfos);
                    return [4 /*yield*/, env_1.validateEnv(typeInfo)];
                case 2:
                    _a.sent();
                    includeSchema = options.output & ComposerFilter.Schema;
                    includeTypeInfo = options.output & ComposerFilter.TypeInfo;
                    return [2 /*return*/, {
                            schema: includeSchema ? render_1.renderSchema(typeInfo, true) : undefined,
                            typeInfo: includeTypeInfo ? typeInfo : undefined,
                        }];
            }
        });
    });
}
exports.composeSchema = composeSchema;
function resolveImports(schemas, resolvers) {
    return __awaiter(this, void 0, void 0, function () {
        var typeInfos, schemas_1, schemas_1_1, schema, _a, _b, e_1_1;
        var e_1, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    typeInfos = [];
                    if (schemas.length === 0) {
                        throw Error("No schema provided");
                    }
                    _d.label = 1;
                case 1:
                    _d.trys.push([1, 6, 7, 8]);
                    schemas_1 = __values(schemas), schemas_1_1 = schemas_1.next();
                    _d.label = 2;
                case 2:
                    if (!!schemas_1_1.done) return [3 /*break*/, 5];
                    schema = schemas_1_1.value;
                    _b = (_a = typeInfos).push;
                    return [4 /*yield*/, resolve_1.resolveImportsAndParseSchemas(schema.schema, schema.absolutePath, resolvers)];
                case 3:
                    _b.apply(_a, [_d.sent()]);
                    _d.label = 4;
                case 4:
                    schemas_1_1 = schemas_1.next();
                    return [3 /*break*/, 2];
                case 5: return [3 /*break*/, 8];
                case 6:
                    e_1_1 = _d.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 8];
                case 7:
                    try {
                        if (schemas_1_1 && !schemas_1_1.done && (_c = schemas_1.return)) _c.call(schemas_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                    return [7 /*endfinally*/];
                case 8: return [2 /*return*/, typeInfos];
            }
        });
    });
}
exports.resolveImports = resolveImports;
