// polywrap/build/lib/helpers/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSimpleClient = void 0;
var ens_plugin_js_1 = require("@polywrap/ens-plugin-js");
var ethereum_plugin_js_1 = require("@polywrap/ethereum-plugin-js");
var ipfs_plugin_js_1 = require("@polywrap/ipfs-plugin-js");
var client_js_1 = require("@polywrap/client-js");
function getSimpleClient(config) {
    var ensAddress = config.ensAddress, ethProvider = config.ethProvider, ipfsProvider = config.ipfsProvider;
    var plugins = [];
    if (ensAddress) {
        plugins.push({
            uri: "wrap://ens/ens.polywrap.eth",
            plugin: ens_plugin_js_1.ensPlugin({
                addresses: {
                    testnet: ensAddress,
                },
            }),
        });
    }
    if (ethProvider) {
        plugins.push({
            uri: "wrap://ens/ethereum.polywrap.eth",
            plugin: ethereum_plugin_js_1.ethereumPlugin({
                networks: {
                    testnet: {
                        provider: ethProvider,
                    },
                },
            }),
        });
    }
    if (ipfsProvider) {
        plugins.push({
            uri: "wrap://ens/ipfs.polywrap.eth",
            plugin: ipfs_plugin_js_1.ipfsPlugin({
                provider: ipfsProvider,
                fallbackProviders: client_js_1.defaultIpfsProviders,
            }),
        });
    }
    return new client_js_1.PolywrapClient({ plugins: plugins });
}
exports.getSimpleClient = getSimpleClient;
