// @polywrap/schema-parse/build/transform/methodParentPointers.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.methodParentPointers = void 0;
function methodParentPointers() {
    var visitorStack = [];
    return {
        enter: {
            ModuleDefinition: function (def) {
                visitorStack.push(def);
                return def;
            },
            ImportedModuleDefinition: function (def) {
                visitorStack.push(def);
                return def;
            },
            MethodDefinition: function (def) {
                var parent = visitorStack.length > 0
                    ? visitorStack[visitorStack.length - 1]
                    : undefined;
                return __assign(__assign({}, def), { parent: parent });
            },
        },
        leave: {
            ModuleDefinition: function (def) {
                visitorStack.pop();
                return def;
            },
            ImportedModuleDefinition: function (def) {
                visitorStack.pop();
                return def;
            },
        },
    };
}
exports.methodParentPointers = methodParentPointers;
