// @polywrap/fs-plugin-js/build/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.plugin = exports.filesystemPlugin = exports.FilesystemPlugin = void 0;
var wrap_1 = require("./wrap");
var path_1 = __importDefault(require("path"));
var fs_1 = __importDefault(require("fs"));
var FilesystemPlugin = /** @class */ (function (_super) {
    __extends(FilesystemPlugin, _super);
    function FilesystemPlugin() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FilesystemPlugin.prototype.tryResolveUri = function (input, _client) {
        return __awaiter(this, void 0, void 0, function () {
            var manifestSearchPatterns, manifest, manifestSearchPatterns_1, manifestSearchPatterns_1_1, manifestSearchPattern, manifestPath;
            var e_1, _a;
            return __generator(this, function (_b) {
                if (input.authority !== "fs") {
                    return [2 /*return*/, null];
                }
                manifestSearchPatterns = [
                    "polywrap.json",
                    "polywrap.yaml",
                    "polywrap.yml",
                ];
                try {
                    for (manifestSearchPatterns_1 = __values(manifestSearchPatterns), manifestSearchPatterns_1_1 = manifestSearchPatterns_1.next(); !manifestSearchPatterns_1_1.done; manifestSearchPatterns_1_1 = manifestSearchPatterns_1.next()) {
                        manifestSearchPattern = manifestSearchPatterns_1_1.value;
                        manifestPath = path_1.default.resolve(input.path, manifestSearchPattern);
                        if (fs_1.default.existsSync(manifestPath)) {
                            try {
                                manifest = fs_1.default.readFileSync(manifestPath, "utf-8");
                            }
                            catch (e) {
                                // TODO: logging
                            }
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (manifestSearchPatterns_1_1 && !manifestSearchPatterns_1_1.done && (_a = manifestSearchPatterns_1.return)) _a.call(manifestSearchPatterns_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                if (manifest) {
                    return [2 /*return*/, { uri: null, manifest: manifest }];
                }
                else {
                    // Noting found
                    return [2 /*return*/, { uri: null, manifest: null }];
                }
                return [2 /*return*/];
            });
        });
    };
    FilesystemPlugin.prototype.getFile = function (input, _client) {
        return __awaiter(this, void 0, void 0, function () {
            var e_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, fs_1.default.promises.readFile(input.path)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        e_2 = _a.sent();
                        return [2 /*return*/, null];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return FilesystemPlugin;
}(wrap_1.Module));
exports.FilesystemPlugin = FilesystemPlugin;
exports.filesystemPlugin = function (opts) {
    return {
        factory: function () { return new FilesystemPlugin(opts); },
        manifest: wrap_1.manifest,
    };
};
exports.plugin = exports.filesystemPlugin;
