// @polywrap/schema-parse/build/extract/Blackboard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Blackboard = void 0;
var graphql_1 = require("graphql");
var Blackboard = /** @class */ (function () {
    function Blackboard(_astNode) {
        this._astNode = _astNode;
    }
    Blackboard.prototype.getCustomTypes = function () {
        if (this._customTypes) {
            return this._customTypes;
        }
        var customTypes = [];
        graphql_1.visit(this._astNode, {
            enter: {
                ObjectTypeDefinition: function (node) {
                    customTypes.push({
                        name: node.name.value,
                        type: "object",
                    });
                },
                EnumTypeDefinition: function (node) {
                    customTypes.push({
                        name: node.name.value,
                        type: "enum",
                    });
                },
            },
        });
        this._customTypes = customTypes;
        return this._customTypes;
    };
    return Blackboard;
}());
exports.Blackboard = Blackboard;
