// @polywrap/schema-parse/build/extract/enum-types.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEnumTypesVisitor = void 0;
var typeInfo_1 = require("../typeInfo");
var visitorEnter = function (enumTypes) { return ({
    EnumTypeDefinition: function (node) {
        var e_1, _a;
        var _b;
        // Skip imported types
        if (node.directives &&
            node.directives.findIndex(function (dir) { return dir.name.value === "imported"; }) > -1) {
            return;
        }
        var constants = [];
        if (node.values) {
            try {
                for (var _c = __values(node.values), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var value = _d.value;
                    constants.push(value.name.value);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        var enumType = typeInfo_1.createEnumDefinition({
            type: node.name.value,
            constants: constants,
            comment: (_b = node.description) === null || _b === void 0 ? void 0 : _b.value,
        });
        enumTypes.push(enumType);
    },
}); };
exports.getEnumTypesVisitor = function (typeInfo) { return ({
    enter: visitorEnter(typeInfo.enumTypes),
}); };
