// @polywrap/schema-parse/build/extract/env-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEnvVisitor = void 0;
var typeInfo_1 = require("../typeInfo");
var object_types_utils_1 = require("./utils/object-types-utils");
var visitorEnter = function (envType, state) { return ({
    ObjectTypeDefinition: function (node) {
        var typeName = node.name.value;
        if (typeInfo_1.isEnvType(typeName)) {
            var type = typeInfo_1.createObjectDefinition({ type: typeName });
            if (typeInfo_1.isClientEnvType(typeName)) {
                envType.client = type;
            }
            else {
                envType.sanitized = type;
            }
            state.currentType = type;
        }
    },
    NonNullType: function (_node) {
        state.nonNullType = true;
    },
    NamedType: function (node) {
        object_types_utils_1.extractNamedType(node, state);
    },
    ListType: function (_node) {
        object_types_utils_1.extractListType(state);
    },
    FieldDefinition: function (node) {
        object_types_utils_1.extractFieldDefinition(node, state);
    },
}); };
var visitorLeave = function (state) { return ({
    ObjectTypeDefinition: function (_node) {
        state.currentType = undefined;
    },
    FieldDefinition: function (_node) {
        state.currentProperty = undefined;
    },
    NonNullType: function (_node) {
        state.nonNullType = false;
    },
}); };
function getEnvVisitor(typeInfo) {
    var state = {};
    return {
        enter: visitorEnter(typeInfo.envType, state),
        leave: visitorLeave(state),
    };
}
exports.getEnvVisitor = getEnvVisitor;
