// @polywrap/schema-parse/build/extract/imported-enum-types.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getImportedEnumTypesVisitor = void 0;
var typeInfo_1 = require("../typeInfo");
var imported_types_utils_1 = require("./utils/imported-types-utils");
var visitorEnter = function (importedEnumTypes) { return ({
    EnumTypeDefinition: function (node) {
        var e_1, _a;
        var _b;
        var constants = [];
        var imported = imported_types_utils_1.extractImportedDefinition(node);
        if (!imported) {
            return;
        }
        if (node.values) {
            try {
                for (var _c = __values(node.values), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var value = _d.value;
                    constants.push(value.name.value);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        var enumType = typeInfo_1.createImportedEnumDefinition({
            type: node.name.value,
            constants: constants,
            uri: imported.uri,
            namespace: imported.namespace,
            nativeType: imported.nativeType,
            comment: (_b = node.description) === null || _b === void 0 ? void 0 : _b.value,
        });
        importedEnumTypes.push(enumType);
    },
}); };
exports.getImportedEnumTypesVisitor = function (typeInfo) { return ({
    enter: visitorEnter(typeInfo.importedEnumTypes),
}); };
