// @polywrap/schema-parse/build/extract/object-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getObjectTypesVisitor = void 0;
var typeInfo_1 = require("../typeInfo");
var object_types_utils_1 = require("./utils/object-types-utils");
var visitorEnter = function (objectTypes, state) { return ({
    ObjectTypeDefinition: function (node) {
        var _a, _b;
        var typeName = node.name.value;
        // Skip non-custom types
        if (typeInfo_1.isModuleType(typeName) || typeInfo_1.isEnvType(typeName)) {
            return;
        }
        // Skip imported types
        if (node.directives &&
            node.directives.findIndex(function (dir) { return dir.name.value === "imported"; }) > -1) {
            return;
        }
        // Create a new TypeDefinition
        var type = typeInfo_1.createObjectDefinition({
            type: typeName,
            interfaces: (_a = node.interfaces) === null || _a === void 0 ? void 0 : _a.map(function (x) {
                return typeInfo_1.createInterfaceImplementedDefinition({ type: x.name.value });
            }),
            comment: (_b = node.description) === null || _b === void 0 ? void 0 : _b.value,
        });
        objectTypes.push(type);
        state.currentType = type;
    },
    NonNullType: function (_node) {
        state.nonNullType = true;
    },
    NamedType: function (node) {
        object_types_utils_1.extractNamedType(node, state);
    },
    ListType: function (_node) {
        object_types_utils_1.extractListType(state);
    },
    FieldDefinition: function (node) {
        object_types_utils_1.extractFieldDefinition(node, state);
    },
}); };
var visitorLeave = function (state) { return ({
    ObjectTypeDefinition: function (_node) {
        state.currentType = undefined;
    },
    FieldDefinition: function (_node) {
        state.currentProperty = undefined;
    },
    NonNullType: function (_node) {
        state.nonNullType = false;
    },
}); };
exports.getObjectTypesVisitor = function (typeInfo) {
    var state = {};
    return {
        enter: visitorEnter(typeInfo.objectTypes, state),
        leave: visitorLeave(state),
    };
};
