// polywrap/build/commands/build.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.build = void 0;
var lib_1 = require("../lib");
var gluegun_1 = require("gluegun");
var path_1 = __importDefault(require("path"));
var readline_1 = __importDefault(require("readline"));
var defaultManifestStr = lib_1.defaultPolywrapManifest.join(" | ");
var pathStr = lib_1.intlMsg.commands_build_options_o_path();
exports.build = {
    setup: function (program) {
        program
            .command("build")
            .alias("b")
            .description(lib_1.intlMsg.commands_build_description())
            .option("-m, --manifest-file <" + pathStr + ">", lib_1.intlMsg.commands_build_options_m({
            default: defaultManifestStr,
        }))
            .option("-o, --output-dir <" + pathStr + ">", "" + lib_1.intlMsg.commands_build_options_o())
            .option("-w, --watch", "" + lib_1.intlMsg.commands_build_options_w())
            .option("-v, --verbose", "" + lib_1.intlMsg.commands_build_options_v())
            .action(function (options) { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, run(__assign(__assign({}, options), { manifestFile: lib_1.parseWasmManifestFileOption(options.manifestFile, undefined), outputDir: lib_1.parseBuildOutputDirOption(options.outputDir, undefined) }))];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
    },
};
function run(options) {
    return __awaiter(this, void 0, void 0, function () {
        var watch, verbose, manifestFile, outputDir, project, dockerLock, schemaComposer, compiler, execute, result, keyPressListener_1, watcher_1;
        var _this = this;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    watch = options.watch, verbose = options.verbose, manifestFile = options.manifestFile, outputDir = options.outputDir;
                    // Ensure docker is installed
                    if (!lib_1.isDockerInstalled()) {
                        console.log(lib_1.intlMsg.lib_docker_noInstall());
                        return [2 /*return*/];
                    }
                    project = new lib_1.PolywrapProject({
                        rootDir: path_1.default.dirname(manifestFile),
                        polywrapManifestPath: manifestFile,
                        quiet: !verbose,
                    });
                    return [4 /*yield*/, project.validate()];
                case 1:
                    _a.sent();
                    dockerLock = new lib_1.FileLock(project.getCachePath("build/DOCKER_LOCK"), gluegun_1.print.error);
                    schemaComposer = new lib_1.SchemaComposer({
                        project: project,
                    });
                    compiler = new lib_1.Compiler({
                        project: project,
                        outputDir: outputDir,
                        schemaComposer: schemaComposer,
                    });
                    execute = function () { return __awaiter(_this, void 0, void 0, function () {
                        var result;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    compiler.reset();
                                    return [4 /*yield*/, compiler.compile()];
                                case 1:
                                    result = _a.sent();
                                    if (!result) {
                                        return [2 /*return*/, result];
                                    }
                                    return [2 /*return*/, true];
                            }
                        });
                    }); };
                    if (!!watch) return [3 /*break*/, 5];
                    return [4 /*yield*/, dockerLock.request()];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, execute()];
                case 3:
                    result = _a.sent();
                    return [4 /*yield*/, dockerLock.release()];
                case 4:
                    _a.sent();
                    if (!result) {
                        process.exitCode = 1;
                        return [2 /*return*/];
                    }
                    return [3 /*break*/, 9];
                case 5: 
                // Execute
                return [4 /*yield*/, dockerLock.request()];
                case 6:
                    // Execute
                    _a.sent();
                    return [4 /*yield*/, execute()];
                case 7:
                    _a.sent();
                    return [4 /*yield*/, dockerLock.release()];
                case 8:
                    _a.sent();
                    keyPressListener_1 = function () {
                        // Watch for escape key presses
                        console.log(lib_1.intlMsg.commands_build_keypressListener_watching() + ": " + project.getManifestDir());
                        console.log(lib_1.intlMsg.commands_build_keypressListener_exit());
                        readline_1.default.emitKeypressEvents(process.stdin);
                        process.stdin.on("keypress", function (str, key) { return __awaiter(_this, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        if (!(key.name == "escape" ||
                                            key.name == "q" ||
                                            (key.name == "c" && key.ctrl))) return [3 /*break*/, 3];
                                        return [4 /*yield*/, watcher_1.stop()];
                                    case 1:
                                        _a.sent();
                                        return [4 /*yield*/, dockerLock.release()];
                                    case 2:
                                        _a.sent();
                                        process.kill(process.pid, "SIGINT");
                                        _a.label = 3;
                                    case 3: return [2 /*return*/];
                                }
                            });
                        }); });
                        if (process.stdin.setRawMode) {
                            process.stdin.setRawMode(true);
                        }
                        process.stdin.resume();
                    };
                    keyPressListener_1();
                    watcher_1 = new lib_1.Watcher();
                    watcher_1.start(project.getManifestDir(), {
                        ignored: [outputDir + "/**", project.getManifestDir() + "/**/wrap/**"],
                        ignoreInitial: true,
                        execute: function (events) { return __awaiter(_this, void 0, void 0, function () {
                            var events_1, events_1_1, event_1;
                            var e_1, _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        try {
                                            // Log all of the events encountered
                                            for (events_1 = __values(events), events_1_1 = events_1.next(); !events_1_1.done; events_1_1 = events_1.next()) {
                                                event_1 = events_1_1.value;
                                                console.log(lib_1.watchEventName(event_1.type) + ": " + event_1.path);
                                            }
                                        }
                                        catch (e_1_1) { e_1 = { error: e_1_1 }; }
                                        finally {
                                            try {
                                                if (events_1_1 && !events_1_1.done && (_a = events_1.return)) _a.call(events_1);
                                            }
                                            finally { if (e_1) throw e_1.error; }
                                        }
                                        // Execute the build
                                        return [4 /*yield*/, dockerLock.request()];
                                    case 1:
                                        // Execute the build
                                        _b.sent();
                                        return [4 /*yield*/, execute()];
                                    case 2:
                                        _b.sent();
                                        return [4 /*yield*/, dockerLock.release()];
                                    case 3:
                                        _b.sent();
                                        // Process key presses
                                        keyPressListener_1();
                                        return [2 /*return*/];
                                }
                            });
                        }); },
                    });
                    _a.label = 9;
                case 9:
                    process.exitCode = 0;
                    return [2 /*return*/];
            }
        });
    });
}
