// polywrap/build/commands/plugin.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.plugin = void 0;
var lib_1 = require("../lib");
var schema_compose_1 = require("@polywrap/schema-compose");
var os_js_1 = require("@polywrap/os-js");
var path_1 = __importDefault(require("path"));
var fs_1 = __importDefault(require("fs"));
var defaultPublishDir = "./build";
var defaultCodegenDir = "./wrap";
var pathStr = lib_1.intlMsg.commands_plugin_options_path();
var defaultManifestStr = lib_1.defaultPluginManifest.join(" | ");
var nodeStr = lib_1.intlMsg.commands_plugin_options_i_node();
var addrStr = lib_1.intlMsg.commands_plugin_options_e_address();
exports.plugin = {
    setup: function (program) {
        var pluginCommand = program
            .command("plugin")
            .alias("p")
            .description(lib_1.intlMsg.commands_plugin_description());
        pluginCommand
            .command("codegen")
            .option("-m, --manifest-file <" + pathStr + ">", "" + lib_1.intlMsg.commands_plugin_options_m({
            default: defaultManifestStr,
        }))
            .option("-p, --publish-dir <" + pathStr + ">", "" + lib_1.intlMsg.commands_plugin_options_publish({
            default: defaultPublishDir,
        }))
            .option("-c, --codegen-dir <" + pathStr + ">", "" + lib_1.intlMsg.commands_plugin_options_codegen({
            default: defaultCodegenDir,
        }))
            .option("-i, --ipfs [<" + nodeStr + ">]", "" + lib_1.intlMsg.commands_plugin_options_i())
            .option("-e, --ens [<" + addrStr + ">]", "" + lib_1.intlMsg.commands_plugin_options_e())
            .action(function (options) { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, run(__assign(__assign({}, options), { manifestFile: lib_1.parsePluginManifestFileOption(options.manifestFile, undefined), publishDir: lib_1.parsePluginPublishDirOption(options.publishDir, undefined), codegenDir: lib_1.parsePluginCodegenDirOption(options.codegenDir, undefined) }))];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
    },
};
function run(options) {
    return __awaiter(this, void 0, void 0, function () {
        var ipfs, ens, manifestFile, codegenDir, publishDir, _a, ipfsProvider, ethProvider, ensAddress, project, manifest, schemaComposer, result, codeGenerator, schemas, publishSchemaPath, publishManifestPath;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    ipfs = options.ipfs, ens = options.ens, manifestFile = options.manifestFile, codegenDir = options.codegenDir, publishDir = options.publishDir;
                    return [4 /*yield*/, lib_1.getTestEnvProviders(ipfs)];
                case 1:
                    _a = _b.sent(), ipfsProvider = _a.ipfsProvider, ethProvider = _a.ethProvider;
                    ensAddress = ens;
                    project = new lib_1.PluginProject({
                        rootDir: path_1.default.dirname(manifestFile),
                        pluginManifestPath: manifestFile,
                    });
                    return [4 /*yield*/, project.validate()];
                case 2:
                    _b.sent();
                    return [4 /*yield*/, project.getManifest()];
                case 3:
                    manifest = _b.sent();
                    schemaComposer = new lib_1.SchemaComposer({
                        project: project,
                        ipfsProvider: ipfsProvider,
                        ethProvider: ethProvider,
                        ensAddress: ensAddress,
                    });
                    result = false;
                    codeGenerator = new lib_1.CodeGenerator({
                        project: project,
                        schemaComposer: schemaComposer,
                        codegenDirAbs: codegenDir,
                    });
                    return [4 /*yield*/, codeGenerator.generate()];
                case 4:
                    result = _b.sent();
                    if (result) {
                        process.exitCode = 0;
                    }
                    else {
                        process.exitCode = 1;
                    }
                    return [4 /*yield*/, schemaComposer.getComposedSchemas(schema_compose_1.ComposerFilter.Schema)];
                case 5:
                    schemas = _b.sent();
                    publishSchemaPath = path_1.default.join(publishDir, "schema.graphql");
                    publishManifestPath = path_1.default.join(publishDir, "polywrap.plugin.json");
                    if (!fs_1.default.existsSync(publishDir)) {
                        fs_1.default.mkdirSync(publishDir);
                    }
                    os_js_1.writeFileSync(publishSchemaPath, schemas.schema);
                    return [4 /*yield*/, lib_1.outputManifest(manifest, publishManifestPath)];
                case 6:
                    _b.sent();
                    return [2 /*return*/];
            }
        });
    });
}
