// @opentelemetry/exporter-zipkin/build/esm/transform.js
/*
 * Copyright The OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var _a;
import * as api from '@opentelemetry/api';
import { hrTimeToMicroseconds } from '@opentelemetry/core';
import * as zipkinTypes from './types';
var ZIPKIN_SPAN_KIND_MAPPING = (_a = {},
    _a[api.SpanKind.CLIENT] = zipkinTypes.SpanKind.CLIENT,
    _a[api.SpanKind.SERVER] = zipkinTypes.SpanKind.SERVER,
    _a[api.SpanKind.CONSUMER] = zipkinTypes.SpanKind.CONSUMER,
    _a[api.SpanKind.PRODUCER] = zipkinTypes.SpanKind.PRODUCER,
    // When absent, the span is local.
    _a[api.SpanKind.INTERNAL] = undefined,
    _a);
export var statusCodeTagName = 'ot.status_code';
export var statusDescriptionTagName = 'ot.status_description';
/**
 * Translate OpenTelemetry ReadableSpan to ZipkinSpan format
 * @param span Span to be translated
 */
export function toZipkinSpan(span, serviceName, statusCodeTagName, statusDescriptionTagName) {
    var zipkinSpan = {
        traceId: span.spanContext().traceId,
        parentId: span.parentSpanId,
        name: span.name,
        id: span.spanContext().spanId,
        kind: ZIPKIN_SPAN_KIND_MAPPING[span.kind],
        timestamp: hrTimeToMicroseconds(span.startTime),
        duration: hrTimeToMicroseconds(span.duration),
        localEndpoint: { serviceName: serviceName },
        tags: _toZipkinTags(span.attributes, span.status, statusCodeTagName, statusDescriptionTagName, span.resource),
        annotations: span.events.length
            ? _toZipkinAnnotations(span.events)
            : undefined,
    };
    return zipkinSpan;
}
/** Converts OpenTelemetry SpanAttributes and SpanStatus to Zipkin Tags format. */
export function _toZipkinTags(attributes, status, statusCodeTagName, statusDescriptionTagName, resource) {
    var tags = {};
    for (var _i = 0, _a = Object.keys(attributes); _i < _a.length; _i++) {
        var key = _a[_i];
        tags[key] = String(attributes[key]);
    }
    tags[statusCodeTagName] = String(api.SpanStatusCode[status.code]);
    if (status.message) {
        tags[statusDescriptionTagName] = status.message;
    }
    Object.keys(resource.attributes).forEach(function (name) { return (tags[name] = String(resource.attributes[name])); });
    return tags;
}
/**
 * Converts OpenTelemetry Events to Zipkin Annotations format.
 */
export function _toZipkinAnnotations(events) {
    return events.map(function (event) { return ({
        timestamp: hrTimeToMicroseconds(event.time),
        value: event.name,
    }); });
}
