// @polywrap/core-js/build/utils/Queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Queue = void 0;
var Queue = /** @class */ (function () {
    function Queue() {
        this.elements = {};
        this.head = 0;
        this.tail = 0;
    }
    Queue.prototype.enqueue = function (element) {
        this.elements[this.tail] = element;
        this.tail++;
    };
    Queue.prototype.dequeue = function () {
        var item = this.elements[this.head];
        if (item === undefined) {
            return undefined;
        }
        delete this.elements[this.head];
        this.head++;
        return item;
    };
    Queue.prototype.peek = function () {
        return this.elements[this.head];
    };
    Object.defineProperty(Queue.prototype, "length", {
        get: function () {
            return this.tail - this.head;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Queue.prototype, "isEmpty", {
        get: function () {
            return this.length === 0;
        },
        enumerable: false,
        configurable: true
    });
    Queue.prototype.toArray = function () {
        var array = [];
        for (var i = this.head; i < this.tail; i++) {
            array.push(this.elements[i]);
        }
        return array;
    };
    return Queue;
}());
exports.Queue = Queue;
