// @polywrap/uts46-plugin-js/build/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.plugin = exports.uts46Plugin = exports.Uts46Plugin = void 0;
/* eslint-disable @typescript-eslint/no-empty-interface */
var wrap_1 = require("./wrap");
// eslint-disable-next-line @typescript-eslint/no-require-imports,@typescript-eslint/no-var-requires
var uts46 = require("idna-uts46-hx/uts46bundle.js");
var Uts46Plugin = /** @class */ (function (_super) {
    __extends(Uts46Plugin, _super);
    function Uts46Plugin() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Uts46Plugin.prototype.toAscii = function (input) {
        return uts46.toAscii(input.value);
    };
    Uts46Plugin.prototype.toUnicode = function (input) {
        return uts46.toUnicode(input.value);
    };
    Uts46Plugin.prototype.convert = function (input) {
        return uts46.convert(input.value);
    };
    return Uts46Plugin;
}(wrap_1.Module));
exports.Uts46Plugin = Uts46Plugin;
exports.uts46Plugin = function (opts) {
    return {
        factory: function () { return new Uts46Plugin(opts); },
        manifest: wrap_1.manifest,
    };
};
exports.plugin = exports.uts46Plugin;
