// @polywrap/schema-parse/build/typeInfo/definitions.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEnvDefinition = exports.createImportedObjectDefinition = exports.createImportedModuleDefinition = exports.createInterfaceDefinition = exports.createCapability = exports.capabilityTypes = exports.createImportedEnumDefinition = exports.createModuleDefinition = exports.createMethodDefinition = exports.createObjectPropertyDefinition = exports.createEnumPropertyDefinition = exports.createScalarPropertyDefinition = exports.createMapPropertyDefinition = exports.createArrayPropertyDefinition = exports.createInterfaceImplementedDefinition = exports.createPropertyDefinition = exports.createArrayDefinition = exports.createMapDefinition = exports.createUnresolvedObjectOrEnumRef = exports.createEnumRef = exports.createEnumDefinition = exports.createScalarDefinition = exports.createMapKeyDefinition = exports.createAnyDefinition = exports.createObjectRef = exports.createObjectDefinition = exports.createGenericDefinition = exports.isKind = exports.DefinitionKind = void 0;
var scalar_1 = require("./scalar");
var module_1 = require("./module");
var map_1 = require("./map");
var DefinitionKind;
(function (DefinitionKind) {
    DefinitionKind[DefinitionKind["Generic"] = 0] = "Generic";
    DefinitionKind[DefinitionKind["Object"] = 1] = "Object";
    DefinitionKind[DefinitionKind["Any"] = 2] = "Any";
    DefinitionKind[DefinitionKind["Scalar"] = 4] = "Scalar";
    DefinitionKind[DefinitionKind["Enum"] = 8] = "Enum";
    DefinitionKind[DefinitionKind["Array"] = 18] = "Array";
    DefinitionKind[DefinitionKind["Property"] = 34] = "Property";
    DefinitionKind[DefinitionKind["Method"] = 64] = "Method";
    DefinitionKind[DefinitionKind["Module"] = 128] = "Module";
    DefinitionKind[DefinitionKind["ImportedModule"] = 256] = "ImportedModule";
    DefinitionKind[DefinitionKind["ImportedEnum"] = 520] = "ImportedEnum";
    DefinitionKind[DefinitionKind["ImportedObject"] = 1025] = "ImportedObject";
    DefinitionKind[DefinitionKind["InterfaceImplemented"] = 2048] = "InterfaceImplemented";
    DefinitionKind[DefinitionKind["UnresolvedObjectOrEnum"] = 4096] = "UnresolvedObjectOrEnum";
    DefinitionKind[DefinitionKind["ObjectRef"] = 8192] = "ObjectRef";
    DefinitionKind[DefinitionKind["EnumRef"] = 16384] = "EnumRef";
    DefinitionKind[DefinitionKind["Interface"] = 32768] = "Interface";
    DefinitionKind[DefinitionKind["Env"] = 65536] = "Env";
    DefinitionKind[DefinitionKind["MapKey"] = 131072] = "MapKey";
    DefinitionKind[DefinitionKind["Map"] = 262146] = "Map";
})(DefinitionKind = exports.DefinitionKind || (exports.DefinitionKind = {}));
function isKind(type, kind) {
    return (type.kind & kind) === kind;
}
exports.isKind = isKind;
function createGenericDefinition(args) {
    return {
        type: args.type,
        name: args.name ? args.name : null,
        required: args.required ? args.required : null,
        kind: DefinitionKind.Generic,
    };
}
exports.createGenericDefinition = createGenericDefinition;
function createObjectDefinition(args) {
    return __assign(__assign({}, createGenericDefinition(args)), { properties: args.properties ? args.properties : [], interfaces: args.interfaces ? args.interfaces : [], comment: args.comment, kind: DefinitionKind.Object });
}
exports.createObjectDefinition = createObjectDefinition;
function createObjectRef(args) {
    return __assign(__assign({}, createGenericDefinition(args)), { kind: DefinitionKind.ObjectRef });
}
exports.createObjectRef = createObjectRef;
function createAnyDefinition(args) {
    return __assign(__assign({}, createGenericDefinition(args)), { array: args.array ? args.array : null, map: args.map ? args.map : null, scalar: args.scalar ? args.scalar : null, object: args.object ? args.object : null, enum: args.enum ? args.enum : null, unresolvedObjectOrEnum: null, kind: DefinitionKind.Any });
}
exports.createAnyDefinition = createAnyDefinition;
function createMapKeyDefinition(args) {
    if (!map_1.isMapKeyType(args.type)) {
        throw Error("createMapKeyDefinition: Unrecognized Map key type provided \"" + args.type + "\"");
    }
    return __assign(__assign({}, createGenericDefinition(args)), { type: args.type, kind: DefinitionKind.Scalar });
}
exports.createMapKeyDefinition = createMapKeyDefinition;
function createScalarDefinition(args) {
    if (!scalar_1.isScalarType(args.type)) {
        throw Error("createScalarDefinition: Unrecognized scalar type provided \"" + args.type + "\"");
    }
    return __assign(__assign({}, createGenericDefinition(args)), { type: args.type, kind: DefinitionKind.Scalar });
}
exports.createScalarDefinition = createScalarDefinition;
function createEnumDefinition(args) {
    return __assign(__assign({}, createGenericDefinition(args)), { type: args.type, kind: DefinitionKind.Enum, constants: args.constants ? args.constants : [], comment: args.comment });
}
exports.createEnumDefinition = createEnumDefinition;
function createEnumRef(args) {
    return __assign(__assign({}, createGenericDefinition(args)), { kind: DefinitionKind.EnumRef });
}
exports.createEnumRef = createEnumRef;
function createUnresolvedObjectOrEnumRef(args) {
    return __assign(__assign({}, createGenericDefinition(args)), { type: args.type, kind: DefinitionKind.UnresolvedObjectOrEnum });
}
exports.createUnresolvedObjectOrEnumRef = createUnresolvedObjectOrEnumRef;
function createMapDefinition(args) {
    return __assign(__assign({}, createAnyDefinition(__assign(__assign({}, args), { array: args.value && isKind(args.value, DefinitionKind.Array)
            ? args.value
            : undefined, map: args.value && isKind(args.value, DefinitionKind.Map)
            ? args.value
            : undefined, scalar: args.value && isKind(args.value, DefinitionKind.Scalar)
            ? args.value
            : undefined, object: args.value && isKind(args.value, DefinitionKind.ObjectRef)
            ? args.value
            : undefined, enum: args.value && isKind(args.value, DefinitionKind.EnumRef)
            ? args.value
            : undefined }))), { key: args.key ? args.key : null, value: args.value ? args.value : null, kind: DefinitionKind.Map });
}
exports.createMapDefinition = createMapDefinition;
function createArrayDefinition(args) {
    return __assign(__assign({}, createAnyDefinition(__assign(__assign({}, args), { array: args.item && isKind(args.item, DefinitionKind.Array)
            ? args.item
            : undefined, map: args.item && isKind(args.item, DefinitionKind.Map)
            ? args.item
            : undefined, scalar: args.item && isKind(args.item, DefinitionKind.Scalar)
            ? args.item
            : undefined, object: args.item && isKind(args.item, DefinitionKind.ObjectRef)
            ? args.item
            : undefined, enum: args.item && isKind(args.item, DefinitionKind.EnumRef)
            ? args.item
            : undefined }))), { item: args.item ? args.item : null, kind: DefinitionKind.Array });
}
exports.createArrayDefinition = createArrayDefinition;
function createPropertyDefinition(args) {
    return __assign(__assign({}, createAnyDefinition(args)), { comment: args.comment, kind: DefinitionKind.Property });
}
exports.createPropertyDefinition = createPropertyDefinition;
function createInterfaceImplementedDefinition(args) {
    return __assign(__assign({}, createAnyDefinition(args)), { kind: DefinitionKind.InterfaceImplemented });
}
exports.createInterfaceImplementedDefinition = createInterfaceImplementedDefinition;
function createArrayPropertyDefinition(args) {
    return createPropertyDefinition(__assign(__assign({}, args), { array: createArrayDefinition(args) }));
}
exports.createArrayPropertyDefinition = createArrayPropertyDefinition;
function createMapPropertyDefinition(args) {
    return createPropertyDefinition(__assign(__assign({}, args), { map: createMapDefinition(args) }));
}
exports.createMapPropertyDefinition = createMapPropertyDefinition;
function createScalarPropertyDefinition(args) {
    return createPropertyDefinition(__assign(__assign({}, args), { scalar: createScalarDefinition(args) }));
}
exports.createScalarPropertyDefinition = createScalarPropertyDefinition;
function createEnumPropertyDefinition(args) {
    return createPropertyDefinition(__assign(__assign({}, args), { enum: createEnumRef(args) }));
}
exports.createEnumPropertyDefinition = createEnumPropertyDefinition;
function createObjectPropertyDefinition(args) {
    return createPropertyDefinition(__assign(__assign({}, args), { object: createObjectRef(args) }));
}
exports.createObjectPropertyDefinition = createObjectPropertyDefinition;
function createMethodDefinition(args) {
    return __assign(__assign({}, createGenericDefinition(__assign(__assign({}, args), { type: "Method" }))), { required: true, arguments: args.arguments ? args.arguments : [], return: args.return, comment: args.comment, kind: DefinitionKind.Method });
}
exports.createMethodDefinition = createMethodDefinition;
function createModuleDefinition(args) {
    return __assign(__assign({}, createGenericDefinition(__assign(__assign({}, args), { type: "Module" }))), { methods: [], imports: args.imports ? args.imports : [], interfaces: args.interfaces ? args.interfaces : [], comment: args.comment, kind: DefinitionKind.Module });
}
exports.createModuleDefinition = createModuleDefinition;
function createImportedEnumDefinition(args) {
    return __assign(__assign({}, createEnumDefinition(args)), { uri: args.uri, namespace: args.namespace, nativeType: args.nativeType, comment: args.comment, kind: DefinitionKind.ImportedEnum });
}
exports.createImportedEnumDefinition = createImportedEnumDefinition;
exports.capabilityTypes = ["getImplementations"];
function createCapability(args) {
    var _a;
    return _a = {},
        _a[args.type] = {
            enabled: args.enabled,
        },
        _a;
}
exports.createCapability = createCapability;
function createInterfaceDefinition(args) {
    return __assign(__assign({}, createGenericDefinition(args)), { namespace: args.namespace, uri: args.uri, nativeType: "Interface", capabilities: args.capabilities, kind: DefinitionKind.Interface });
}
exports.createInterfaceDefinition = createInterfaceDefinition;
function createImportedModuleDefinition(args) {
    if (!module_1.isModuleType(args.nativeType)) {
        throw Error("createImportedModuleDefinition: Unrecognized module type provided \"" + args.nativeType + "\"");
    }
    return __assign(__assign({}, createGenericDefinition(__assign(__assign({}, args), { type: args.namespace + "_" + args.nativeType }))), { methods: [], uri: args.uri, namespace: args.namespace, nativeType: args.nativeType, comment: args.comment, isInterface: args.isInterface, kind: DefinitionKind.ImportedModule });
}
exports.createImportedModuleDefinition = createImportedModuleDefinition;
function createImportedObjectDefinition(args) {
    return __assign(__assign({}, createObjectDefinition(args)), { uri: args.uri, namespace: args.namespace, nativeType: args.nativeType, comment: args.comment, kind: DefinitionKind.ImportedObject });
}
exports.createImportedObjectDefinition = createImportedObjectDefinition;
function createEnvDefinition(args) {
    return {
        kind: DefinitionKind.Env,
        sanitized: args.sanitized,
        client: args.client,
    };
}
exports.createEnvDefinition = createEnvDefinition;
