// @polywrap/graph-node-plugin-js/build/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.plugin = exports.graphNodePlugin = exports.GraphNodePlugin = void 0;
var wrap_1 = require("./wrap");
var GraphNodePlugin = /** @class */ (function (_super) {
    __extends(GraphNodePlugin, _super);
    function GraphNodePlugin(config) {
        return _super.call(this, config) || this;
    }
    GraphNodePlugin.prototype.querySubgraph = function (input, client) {
        return __awaiter(this, void 0, void 0, function () {
            var subgraphAuthor, subgraphName, query, _a, data, error, responseJson, responseErrors;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        subgraphAuthor = input.subgraphAuthor, subgraphName = input.subgraphName, query = input.query;
                        return [4 /*yield*/, wrap_1.HTTP_Module.post({
                                url: this.config.provider + "/subgraphs/name/" + subgraphAuthor + "/" + subgraphName,
                                request: {
                                    body: JSON.stringify({
                                        query: query,
                                    }),
                                    responseType: "TEXT",
                                },
                            }, client)];
                    case 1:
                        _a = _b.sent(), data = _a.data, error = _a.error;
                        if (error) {
                            throw new Error("GraphNodePlugin: errors encountered. Error: " + error);
                        }
                        if (!data) {
                            throw new Error("GraphNodePlugin: data is undefined.");
                        }
                        if (!data.body) {
                            throw Error("GraphNodePlugin: body is undefined.");
                        }
                        responseJson = JSON.parse(data.body);
                        responseErrors = responseJson.errors;
                        if (responseErrors) {
                            throw new Error("GraphNodePlugin: errors in query string. Errors:\n        " + responseErrors
                                .map(function (err) {
                                return err.locations
                                    ? "\n -Locations: " + err.locations
                                        .map(function (loc) { return "(col: " + loc.column + ", line: " + loc.line + ")"; })
                                        .join(", ") + " \n-Message: " + err.message
                                    : "\n-Message: " + err.message;
                            })
                                .join("\n") + "\n      ");
                        }
                        return [2 /*return*/, JSON.stringify(responseJson)];
                }
            });
        });
    };
    return GraphNodePlugin;
}(wrap_1.Module));
exports.GraphNodePlugin = GraphNodePlugin;
exports.graphNodePlugin = function (opts) {
    return {
        factory: function () { return new GraphNodePlugin(opts); },
        manifest: wrap_1.manifest,
    };
};
exports.plugin = exports.graphNodePlugin;
