// @polywrap/ethereum-plugin-js/build/utils/mapping.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromAccess = exports.toAccess = exports.fromLog = exports.toLog = exports.fromTxRequest = exports.toTxRequest = exports.toTxResponse = exports.fromTxReceipt = exports.toTxReceipt = void 0;
var ethers_1 = require("ethers");
exports.toTxReceipt = function (receipt) {
    var _a;
    return ({
        to: receipt.to || "",
        from: receipt.from,
        contractAddress: receipt.contractAddress,
        transactionIndex: receipt.transactionIndex,
        root: receipt.root,
        gasUsed: receipt.gasUsed.toString(),
        logsBloom: receipt.logsBloom,
        transactionHash: receipt.transactionHash,
        logs: receipt.logs.map(exports.toLog),
        blockNumber: receipt.blockNumber.toString(),
        blockHash: receipt.blockHash,
        confirmations: receipt.confirmations,
        cumulativeGasUsed: receipt.cumulativeGasUsed.toString(),
        effectiveGasPrice: ((_a = receipt.effectiveGasPrice) === null || _a === void 0 ? void 0 : _a.toString()) || "0",
        byzantium: receipt.byzantium,
        type: receipt.type || 0,
        status: receipt.status,
    });
};
exports.fromTxReceipt = function (receipt) { return ({
    to: receipt.to,
    from: receipt.from,
    contractAddress: receipt.contractAddress,
    transactionIndex: receipt.transactionIndex,
    root: receipt.root || undefined,
    gasUsed: ethers_1.ethers.BigNumber.from(receipt.gasUsed),
    logsBloom: receipt.logsBloom,
    transactionHash: receipt.transactionHash,
    logs: receipt.logs.map(exports.fromLog),
    blockNumber: Number(receipt.blockNumber),
    blockHash: receipt.blockHash,
    confirmations: receipt.confirmations,
    cumulativeGasUsed: ethers_1.ethers.BigNumber.from(receipt.cumulativeGasUsed),
    byzantium: receipt.byzantium,
    status: receipt.status || undefined,
    effectiveGasPrice: ethers_1.ethers.BigNumber.from(receipt.effectiveGasPrice),
    type: receipt.type,
}); };
exports.toTxResponse = function (response) {
    var _a, _b, _c;
    return ({
        hash: response.hash,
        to: response.to,
        from: response.from,
        nonce: response.nonce,
        gasLimit: response.gasLimit.toString(),
        gasPrice: (_a = response.gasPrice) === null || _a === void 0 ? void 0 : _a.toString(),
        data: response.data,
        value: response.value.toString(),
        chainId: response.chainId.toString(),
        blockNumber: (_b = response.blockNumber) === null || _b === void 0 ? void 0 : _b.toString(),
        blockHash: response.blockHash,
        timestamp: response.timestamp,
        confirmations: response.confirmations,
        raw: response.raw,
        r: response.r,
        s: response.s,
        v: response.v,
        type: response.type || undefined,
        accessList: (_c = response.accessList) === null || _c === void 0 ? void 0 : _c.map(exports.toAccess),
    });
};
exports.toTxRequest = function (request) {
    var _a, _b, _c, _d, _e;
    return ({
        to: request.to,
        from: request.from,
        nonce: request.nonce ? Number(request.nonce.toString()) : undefined,
        gasLimit: (_a = request.gasLimit) === null || _a === void 0 ? void 0 : _a.toString(),
        gasPrice: (_b = request.gasPrice) === null || _b === void 0 ? void 0 : _b.toString(),
        data: (_c = request.data) === null || _c === void 0 ? void 0 : _c.toString(),
        value: (_d = request.value) === null || _d === void 0 ? void 0 : _d.toString(),
        chainId: (_e = request.chainId) === null || _e === void 0 ? void 0 : _e.toString(),
        type: request.type,
    });
};
exports.fromTxRequest = function (request) { return ({
    to: request.to || undefined,
    from: request.from || undefined,
    nonce: request.nonce || undefined,
    gasLimit: request.gasLimit
        ? ethers_1.ethers.BigNumber.from(request.gasLimit)
        : undefined,
    gasPrice: request.gasPrice
        ? ethers_1.ethers.BigNumber.from(request.gasPrice)
        : undefined,
    data: request.data || undefined,
    value: request.value ? ethers_1.ethers.BigNumber.from(request.value) : undefined,
    chainId: request.chainId ? Number.parseInt(request.chainId) : undefined,
    type: request.type || undefined,
}); };
exports.toLog = function (log) { return ({
    blockNumber: log.blockNumber.toString(),
    blockHash: log.blockHash,
    transactionIndex: log.transactionIndex,
    removed: !!log.removed,
    address: log.address,
    data: log.data,
    topics: log.topics,
    transactionHash: log.transactionHash,
    logIndex: log.logIndex,
}); };
exports.fromLog = function (log) { return ({
    blockNumber: Number(log.blockNumber),
    blockHash: log.blockHash,
    transactionIndex: log.transactionIndex,
    removed: log.removed,
    address: log.address,
    data: log.data,
    topics: log.topics,
    transactionHash: log.transactionHash,
    logIndex: log.logIndex,
}); };
exports.toAccess = function (access) { return ({
    address: access.address,
    storageKeys: access.storageKeys,
}); };
exports.fromAccess = function (access) { return ({
    address: access.address,
    storageKeys: access.storageKeys,
}); };
