// @polywrap/ipfs-plugin-js/build/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.plugin = exports.ipfsPlugin = exports.IpfsPlugin = void 0;
var wrap_1 = require("./wrap");
var exec_1 = require("./utils/exec");
// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
var isIPFS = require("is-ipfs");
// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports, @typescript-eslint/naming-convention
var createIpfsClient = require("@dorgjelli-test/ipfs-http-client-lite");
var getOptions = function (input, env) {
    var options = input || {};
    if (options.disableParallelRequests === undefined ||
        options.disableParallelRequests === null) {
        options.disableParallelRequests = env.disableParallelRequests;
    }
    return options;
};
var IpfsPlugin = /** @class */ (function (_super) {
    __extends(IpfsPlugin, _super);
    function IpfsPlugin(config) {
        var _this = _super.call(this, config) || this;
        _this._ipfs = createIpfsClient(_this.config.provider);
        return _this;
    }
    IpfsPlugin.isCID = function (cid) {
        return isIPFS.cid(cid) || isIPFS.cidPath(cid) || isIPFS.ipfsPath(cid);
    };
    IpfsPlugin.prototype.cat = function (cid, options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._execWithOptions("cat", function (ipfs, _provider, options) {
                            return ipfs.cat(cid, options);
                        }, options)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    IpfsPlugin.prototype.catToString = function (cid, options) {
        return __awaiter(this, void 0, void 0, function () {
            var buffer;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.cat(cid, options)];
                    case 1:
                        buffer = _a.sent();
                        return [2 /*return*/, buffer.toString("utf-8")];
                }
            });
        });
    };
    IpfsPlugin.prototype.catFile = function (input) {
        return __awaiter(this, void 0, void 0, function () {
            var options;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        options = getOptions(input.options, this.env);
                        return [4 /*yield*/, this.cat(input.cid, options)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    IpfsPlugin.prototype.resolve = function (input) {
        return __awaiter(this, void 0, void 0, function () {
            var options;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        options = getOptions(input.options, this.env);
                        return [4 /*yield*/, this._execWithOptions("resolve", function (ipfs, provider, options) { return __awaiter(_this, void 0, void 0, function () {
                                var path;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, ipfs.resolve(input.cid, options)];
                                        case 1:
                                            path = (_a.sent()).path;
                                            return [2 /*return*/, {
                                                    cid: path,
                                                    provider: provider,
                                                }];
                                    }
                                });
                            }); }, options)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    // uri-resolver.core.polywrap.eth
    IpfsPlugin.prototype.tryResolveUri = function (input) {
        return __awaiter(this, void 0, void 0, function () {
            var manifestSearchPatterns, manifest, manifestSearchPatterns_1, manifestSearchPatterns_1_1, manifestSearchPattern, e_1, e_2_1;
            var e_2, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (input.authority !== "ipfs") {
                            return [2 /*return*/, null];
                        }
                        if (!IpfsPlugin.isCID(input.path)) {
                            // Not a valid CID
                            return [2 /*return*/, { manifest: null, uri: null }];
                        }
                        manifestSearchPatterns = [
                            "polywrap.json",
                            "polywrap.yaml",
                            "polywrap.yml",
                        ];
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 8, 9, 10]);
                        manifestSearchPatterns_1 = __values(manifestSearchPatterns), manifestSearchPatterns_1_1 = manifestSearchPatterns_1.next();
                        _b.label = 2;
                    case 2:
                        if (!!manifestSearchPatterns_1_1.done) return [3 /*break*/, 7];
                        manifestSearchPattern = manifestSearchPatterns_1_1.value;
                        _b.label = 3;
                    case 3:
                        _b.trys.push([3, 5, , 6]);
                        return [4 /*yield*/, this.catToString(input.path + "/" + manifestSearchPattern, {
                                timeout: 5000,
                                disableParallelRequests: this.env.disableParallelRequests,
                            })];
                    case 4:
                        manifest = _b.sent();
                        return [3 /*break*/, 6];
                    case 5:
                        e_1 = _b.sent();
                        return [3 /*break*/, 6];
                    case 6:
                        manifestSearchPatterns_1_1 = manifestSearchPatterns_1.next();
                        return [3 /*break*/, 2];
                    case 7: return [3 /*break*/, 10];
                    case 8:
                        e_2_1 = _b.sent();
                        e_2 = { error: e_2_1 };
                        return [3 /*break*/, 10];
                    case 9:
                        try {
                            if (manifestSearchPatterns_1_1 && !manifestSearchPatterns_1_1.done && (_a = manifestSearchPatterns_1.return)) _a.call(manifestSearchPatterns_1);
                        }
                        finally { if (e_2) throw e_2.error; }
                        return [7 /*endfinally*/];
                    case 10:
                        if (manifest) {
                            return [2 /*return*/, { uri: null, manifest: manifest }];
                        }
                        else {
                            // Nothing found
                            return [2 /*return*/, { uri: null, manifest: null }];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    IpfsPlugin.prototype.getFile = function (input) {
        return __awaiter(this, void 0, void 0, function () {
            var result, e_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, this.resolve({
                                cid: input.path,
                                options: {
                                    timeout: 5000,
                                    disableParallelRequests: this.env.disableParallelRequests,
                                },
                            })];
                    case 1:
                        result = _a.sent();
                        if (!result) {
                            return [2 /*return*/, null];
                        }
                        return [4 /*yield*/, this.cat(result.cid, {
                                provider: result.provider,
                                timeout: 20000,
                                disableParallelRequests: true,
                            })];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3:
                        e_3 = _a.sent();
                        return [2 /*return*/, null];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    IpfsPlugin.prototype.addFile = function (input) {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._ipfs.add(new Uint8Array(input.data))];
                    case 1:
                        result = _a.sent();
                        if (result.length === 0) {
                            throw Error("IpfsPlugin:add failed to add contents. Result of length 0 returned.");
                        }
                        return [2 /*return*/, result[0].hash.toString()];
                }
            });
        });
    };
    IpfsPlugin.prototype._execWithOptions = function (operation, func, options) {
        return __awaiter(this, void 0, void 0, function () {
            var timeout, providers, ipfs, defaultProvider;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!options) return [3 /*break*/, 2];
                        return [4 /*yield*/, exec_1.execSimple(operation, this._ipfs, this.config.provider, 0, func)];
                    case 1: 
                    // Default behavior if no options are provided
                    return [2 /*return*/, _a.sent()];
                    case 2:
                        timeout = options.timeout || 0;
                        providers = __spread([
                            this.config.provider
                        ], (this.config.fallbackProviders || []));
                        ipfs = this._ipfs;
                        defaultProvider = this.config.provider;
                        // Use the provider default override specified
                        if (options.provider) {
                            providers = __spread([options.provider], providers);
                            ipfs = createIpfsClient(options.provider);
                            defaultProvider = options.provider;
                        }
                        return [4 /*yield*/, exec_1.execFallbacks(operation, ipfs, defaultProvider, providers, timeout, func, {
                                parallel: !options.disableParallelRequests,
                            })];
                    case 3: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return IpfsPlugin;
}(wrap_1.Module));
exports.IpfsPlugin = IpfsPlugin;
exports.ipfsPlugin = function (opts) {
    return {
        factory: function () { return new IpfsPlugin(opts); },
        manifest: wrap_1.manifest,
    };
};
exports.plugin = exports.ipfsPlugin;
